### PR TITLE
Fix integer overflow check in number parser

### LINKED ISSUE #851: Parsing IPv4 address from a string sometimes works although it should fail
The parser behaves weirdly if an octet has 4 digits starting with a 1 in the range `1137` to `1279`.

Here's a few observations:
* `"1136.12.12.12"` fails to parse
* `"1137.12.12.12"` parses to `113.12.12.12`
*  `"1138.12.12.12"` parses to `114.12.12.12`
*  continues up to `"1279.12.12.12"`  which parses to `255.12.12.12`
*  `"1280.12.12.12"`  and above fails again

And a test case for `test/ipv4_address.hpp`

```cpp
CAF_TEST(bug) {
  auto should_fail = [](std::string addr) {
    ipv4_address x;
    auto err = parse(addr, x);
    CAF_CHECK(err);
  };
  should_fail("1136.12.12.12");
  should_fail("1137.12.12.12");
  should_fail("1279.12.12.12");
  should_fail("1280.12.12.12");
}
```

### PR DESCRIPTION
**Pull request by Neverlord**
Fixes #851.

### CODE REVIEW

File: libcaf_core/caf/detail/parser/add_ascii.hpp [hunk 1]
```diff
@@ -29,11 +33,24 @@ namespace parser {
 // @returns `false` on an overflow, otherwise `true`.
 // @pre `isdigit(c) || (Base == 16 && isxdigit(c))`
 template <int Base, class T>
-bool add_ascii(T& x, char c) {
+bool add_ascii(T& x, char c, enable_if_tt<std::is_integral<T>, int> u = 0) {
+  CAF_IGNORE_UNUSED(u);
+  ascii_to_int<Base, T> f;
+  auto y = f(c);
+  if (x > ((std::numeric_limits<T>::max() / Base) - y)
+            + (std::numeric_limits<T>::max() % Base))
```
Comment (josephnoir) [inline libcaf_core/caf/detail/parser/add_ascii.hpp]: This does not work in all cases. Here are a few test cases that should fail, but don't and give a wrong address as a result:

```
  should_fail("260.12.12.12");
  should_fail("261.12.12.12"); 
  should_fail("262.12.12.12");
  should_fail("263.12.12.12");
  should_fail("264.12.12.12");
  should_fail("270.12.12.12");
  should_fail("271.12.12.12"); 
  should_fail("272.12.12.12"); 
  should_fail("273.12.12.12"); 
  should_fail("280.12.12.12"); 
  should_fail("281.12.12.12");  
  should_fail("282.12.12.12"); 
  should_fail("291.12.12.12"); 
  should_fail("300.12.12.12");
```

File: libcaf_core/caf/detail/parser/read_ipv4_address.hpp [hunk 2]
```diff
@@ -53,7 +53,8 @@ void read_ipv4_octet(state<Iterator, Sentinel>& ps, Consumer& consumer) {
   uint8_t res = 0;
   // Reads the a decimal place.
   auto rd_decimal = [&](char c) {
-    return add_ascii<10>(res, c);
+    auto b = add_ascii<10>(res, c) && res <= 255;
```
Comment (josephnoir) [inline libcaf_core/caf/detail/parser/read_ipv4_address.hpp]: Shouldn't `res <= 255` always be true for `uint8_t`? (And if the check is necessary, shouldn't it use `std::numeric_limits<uint8_t>::max()`.)

File: libcaf_core/caf/detail/parser/add_ascii.hpp (hunk 1 above)
Comment (josephnoir) [reply libcaf_core/caf/detail/parser/add_ascii.hpp]: Oh, and why is the `BASE` part of the check. It is only important for converting the ascii value to an integer.

Comment (josephnoir) [reply libcaf_core/caf/detail/parser/add_ascii.hpp]: Ok, I see why the BASE is necessary here. Sorry for the spam. Still doesn't work though.

Comment (josephnoir) [reply libcaf_core/caf/detail/parser/add_ascii.hpp]: Adding a check to see if we can even add another digit would address this:
```cpp  
template <int Base, class T>      
bool add_ascii(T& x, char c, enable_if_tt<std::is_integral<T>, int> u = 0) {      
  CAF_IGNORE_UNUSED(u);      
  // Check if we can add another digit.      
  T res = (x * Base);                                          
  if (res / Base != x)                                         
    return false;                                           
  // Check if we can add this specific digit.              
  ascii_to_int<Base, T> f;                                                                
  auto y = f(c);                                                                                  
  if (x > ((std::numeric_limits<T>::max() / Base) - y)                                                           
            + (std::numeric_limits<T>::max() % Base))                                                            
    return false;                                                                                                
  x = res + y;                                                                                                   
  return true;                                                                                                   
}
```

**Comment by Neverlord** [comment]
Thanks for double-checking my bad code! The new bound checks in `add_ascii` and `sub_ascii` should do the trick - and are reasonably tested now.

File: libcaf_core/caf/detail/parser/add_ascii.hpp [hunk 3]
```diff
@@ -29,11 +32,26 @@ namespace parser {
 // @returns `false` on an overflow, otherwise `true`.
 // @pre `isdigit(c) || (Base == 16 && isxdigit(c))`
 template <int Base, class T>
-bool add_ascii(T& x, char c) {
+bool add_ascii(T& x, char c, enable_if_tt<std::is_integral<T>, int> u = 0) {
+  CAF_IGNORE_UNUSED(u);
+  if (x > (std::numeric_limits<T>::max() / Base))
+    return false;
+  x *= Base;
```
Comment (josephnoir) [inline libcaf_core/caf/detail/parser/add_ascii.hpp]: If parsing fails this mutates the input value `x`.

File: libcaf_core/caf/detail/parser/sub_ascii.hpp [hunk 4]
```diff
@@ -29,11 +32,26 @@ namespace parser {
 // @returns `false` on an underflow, otherwise `true`.
 // @pre `isdigit(c) || (Base == 16 && isxdigit(c))`
 template <int Base, class T>
-bool sub_ascii(T& x, char c) {
+bool sub_ascii(T& x, char c, enable_if_tt<std::is_integral<T>, int> u = 0) {
+  CAF_IGNORE_UNUSED(u);
+  if (x < (std::numeric_limits<T>::min() / Base))
+    return false;
+  x *= Base;
```
Comment (josephnoir) [inline libcaf_core/caf/detail/parser/sub_ascii.hpp]: Same as above.

**Comment by Neverlord** [comment]
I think it's never safe to rely on parser output variables in case of an error. However, documenting this fact does not hurt either.

**Pull request merged**